src/hotspot/share/prims/jvmtiTagMapTable.cpp

```diff
@@ -38,189 +38,109 @@
+}
 
-      // Peek the object to check if it is the right target.
-      oop target = p->object_no_keepalive();
+JvmtiTagMapTable::~JvmtiTagMapTable() {
+  clear();
+}
 
-      // The obj is in the table as a target already
-      if (target == obj) {
-        ResourceMark rm;
-        log_trace(jvmti, table)("JvmtiTagMap entry found for %s index %d",
-                                obj->print_value_string(), index);
-        // The object() accessor makes sure the target object is kept alive before
-        // leaking out.
-        (void)p->object();
-        return p;
-      }
-    }
+jlong JvmtiTagMapTable::find(oop obj) {
+  if (is_empty()) {
+    return 0;
   }
-  return NULL;
-}
 
-JvmtiTagMapEntry* JvmtiTagMapTable::find(oop obj) {
   if (obj->fast_no_hash_check()) {
-    // Objects in the table all have a hashcode.
-    return NULL;
+    return 0;
   }
-  unsigned int hash = compute_hash(obj);
-  int index = hash_to_index(hash);
-  return find(index, hash, obj);
-}
-
-JvmtiTagMapEntry* JvmtiTagMapTable::add(oop obj, jlong tag) {
-  unsigned int hash = compute_hash(obj);
-  int index = hash_to_index(hash);
-  // One was added while acquiring the lock
-  assert(find(index, hash, obj) == NULL, "shouldn't already be present");
-
-  // obj was read with AS_NO_KEEPALIVE, or equivalent.
-  // The object needs to be kept alive when it is published.
-  Universe::heap()->keep_alive(obj);
 
-  WeakHandle w(JvmtiExport::weak_tag_storage(), obj);
-  JvmtiTagMapEntry* p = new_entry(hash, w, tag);
-  Hashtable<WeakHandle, mtServiceability>::add_entry(index, p);
-  ResourceMark rm;
-  log_trace(jvmti, table)("JvmtiTagMap entry added for %s index %d",
-                          obj->print_value_string(), index);
-
-  // Resize if the table is getting too big.
-  resize_if_needed();
+  JvmtiTagMapKey jtme(obj);
+  jlong* found = _table.get(jtme);
+  return found == NULL ? 0 : *found;
+}
 
-  return p;
+bool JvmtiTagMapTable::add(oop obj, jlong tag) {
+  JvmtiTagMapKey new_entry(obj);
+  bool is_added = false;
+  _table.put_if_absent(new_entry, tag, &is_added);
+  assert(is_added, "should be added");
+  return is_added;
 }
 
-void JvmtiTagMapTable::remove(oop obj) {
-  unsigned int hash = compute_hash(obj);
-  int index = hash_to_index(hash);
-  JvmtiTagMapEntry** p = bucket_addr(index);
-  JvmtiTagMapEntry* entry = bucket(index);
-  while (entry != NULL) {
-    oop target = entry->object_no_keepalive();
-    if (target != NULL && target == obj) {
-      log_trace(jvmti, table)("JvmtiTagMap entry removed for index %d", index);
-      *p = entry->next();
-      free_entry(entry);
-      return; // done
-    }
-    // get next entry and address
-    p = entry->next_addr();
-    entry = entry->next();
-  }
+bool JvmtiTagMapTable::update(oop obj, jlong tag) {
+  JvmtiTagMapKey new_entry(obj);
+  bool is_updated = _table.put(new_entry, tag) == false;
+  assert(is_updated, "should be updated and not added");
```

Comment:
This API is confusing. You have `add` and `replace` but in product mode an `add` can `replace` and a `replace` can `add`. It isn't clear if these should be allowed rare conditions or fatal errors.

Reply:
I agree that this is ambigious. The `jvmtiTagMap` calls `add/update` methods of `jvmtiTagMapTable` which in turn calls `resourceHashTable` methods `put` and `put_if_absent`.
`put` and `put_if_absent` can be used for both adding and updating and reporting it in returned values.
`jvmtiTagMap` calls to `jvmtiTagMapTable` add/update  DO NOT CARE about the returned values. For these calls, it doesn't mater if an add (update) method resulted in an update (add). 
So which one of the following cases would be correct?
 - Based on the `jvmtiTagMap` calls, let the add/update be void and ignore the differences of what really happens in the table. Or
 - Based on the `resourceHashTable` bool-valued `put` and `put_if_absent` methods, let the add/update be bool-values and leave the decision on "what really happened in the table" to the callers in `jvmtiTagMap`. (current implementation)

Reply:
The issue is not the underlying RHT methods but the semantics of the `jvmtiTagMap` methods. If a call to add always expects to add then it should be a fatal error if it actually did an update as that indicates something is broken. Similarly if an update actually does an add.

Reply:
The JvmtiTagMap code adds/updates and removes the entries like below in two places which could probably be simplified, but I think that's outside the scope of this RFE.   I suggest removing the bool return from add, remove and update, and add the assert(true) in the remove case.  There's already an assert that add/update happened in the other cases.

 ``` // if the object is not already tagged then we tag it
  if (found_tag == 0) {
    if (tag != 0) {
      hashmap->add(o, tag);
    } else {
      // no-op
    }
  } else {
    // if the object is already tagged then we either update
    // the tag (if a new tag value has been provided)
    // or remove the object if the new tag value is 0.
    if (tag == 0) {
      hashmap->remove(o);
    } else {
      hashmap->update(o, tag);
    }
  }
```

Reply:
The methods of HashmapTable are void and have assert to verify the expected function (add/update/remove) is done.

Reply:
I'm still somewhat concerned about the use of this API and how it will behave in product mode, but okay ...